plugins/newsdownloader.koplugin/epubdownloadbackend.lua

```diff
@@ -24,6 +24,47 @@ local max_redirects = 5; --prevent infinite redirects
 local TIMEOUT_CODE = "timeout" -- from socket.lua
 local MAXTIME_CODE = "maxtime reached" -- from sink_table_with_maxtime
 
+local function filter(text, element)
+    local htmlparser = require("htmlparser")
+    local root = htmlparser.parse(text)
+    local filtered = ""
+    
+    local selectors = {
+        "main",
+        "article",
+        "div#main",
+        "#main-article",
+        ".main-content",
+        "#body",
+        "#content",
+        ".content",
+        "div#article",
+        "div.article",
+        "div.post",
+        "div.post-outer",
+        ".l-root",
+        ".content-container",
+        ".StandardArticleBody_body",
+        "div#article-inner",
+        "div#newsstorytext",
+        "div.general",
+        }
+    
+    if element then selectors[0] = element end
+    
+    for _,sel in pairs(selectors) do
+       local elements = root:select(sel)
+       if not elements then return text end
+       for _,e in ipairs(elements) do
+           filtered = e:getcontent()
+           if filtered then goto endloop end
+       end
+end
+
+::endloop::
+return "<!DOCTYPE html><html><head></head><body>" .. filtered .. "</body></html>"
+end
```

Comment:
You can avoid that (ugly) goto by doing:
```lua
    for _, sel in pairs(selectors) do
       local elements = root:select(sel)
       if not elements then return text end
       local stop = false
       for _, e in ipairs(elements) do
           filtered = e:getcontent()
           if filtered then
               stop = true
               break
           end
       end
       if stop then
           break
       end
    end
    return "<!DOCTYPE html><html><head></head><body>" .. filtered .. "</body></html>"
```

Try to keep if/end aligned (your indentation of the `end` at line 62 was not aligned to its `for`, which made it strange to read :) - and a space after comma (`for _, sel`  rather than `for _,sel `) to make that easier to read.
Looks fine otherwise.

Reply:
ok, i'll clean it up, thanks for a cleaner solution